agents/dqn.py

```diff
@@ -1,24 +1,132 @@
+            if 'path_memory' in config else None,
+            node_idx=config.get('node_idx', None),
+            sample_size=config.get('sample_size', 100)
+        )
+
+    def forward(self, inputs: Dict[str, Any]) -> Dict[str, Any]:
+        # Shape: [batch_size]
+        current_node_idx = inputs[self._current_node_idx_prefix]
+        # Shape: [batch_size]
+        destination_node_idx = inputs[self._destination_node_idx_prefix]
+        # TensorWithMask
+        neighbor_node_ids = inputs[self._neighbors_node_ids_prefix]
+
+        next_neighbor_q = self._q_network(
+            current_node_idx=current_node_idx,
+            neighbor_node_ids=neighbor_node_ids,
+            destination_node_idx=destination_node_idx
         )
 
-    def forward(self, inputs):
-        for part in self._parts:
-            inputs = part(inputs)
-        return inputs
+        next_neighbor_ids = neighbor_node_ids.flatten_values[torch.argmax(next_neighbor_q, dim=1)]
+        next_neighbors_idx = _with_random_research(
+            next_neighbor_ids,
+            neighbor_node_ids,
+            self._research_prob
+        )
+
+        if self._bag_trajectory_memory is not None:
+            # TODO[Vladimir Baikalov]: Think about how to generalize
+            bag_ids = inputs.get(self._bag_idx_prefix, None)
+
+            self._bag_trajectory_memory.add_to_trajectory(
+                bag_ids=bag_ids,
+                node_idxs=current_node_idx,
+                extra_infos=zip(
+                    current_node_idx,
+                    neighbor_node_ids,
+                    destination_node_idx,
+                    next_neighbors_idx
+                )
+            )
+        return {
+            'predicted_next_node_idx': next_neighbors_idx,
+            'predicted_next_node_q': next_neighbor_q,
+        }
+
+    def learn(self):
+        for trajectory in self._bag_trajectory_memory.sample_trajectories_for_node_idx(
+                self._node_idx,
+                self._trajectory_sample_size
+        ):
+            target = trajectory[0]['reward']
+            current_node_idx, neighbor_node_ids, destination_node_idx, next_neighbor = trajectory[0]['extra_infos']
+            if len(trajectory) > 0:
+                current_node_idx_, neighbor_node_ids_, destination_node_idx_, _ = trajectory[1]['extra_infos']
+                neighbor_q_ = self._q_network(
+                    current_node_idx=current_node_idx_,
+                    neighbor_node_ids=neighbor_node_ids_,
+                    destination_node_idx=destination_node_idx_
+                )
+                target += self._discount_factor * neighbor_q_
+            neighbor_q = self._q_network(
+                current_node_idx=current_node_idx,
+                neighbor_node_ids=neighbor_node_ids,
+                destination_node_idx=destination_node_idx
+            )
+            loss = (target.detach() - neighbor_q[next_neighbor]) ** 2
```

Comment:
This is incorrect! By using `detach` you remove all the information about gradient flow. Please, change it, otherwise `loss.backward()` will not work properly. Description of a `detach` method: "Returns a new Tensor, detached from the current graph. The result will never require gradient."

Reply:
I know what detach does) I thought that old value should be detached in dqn, may be discus later in telegram?